src/librbd/ImageWatcher.cc

```diff
@@ -403,20 +403,22 @@ template <typename I>
 void ImageWatcher<I>::notify_quiesce(const AsyncRequestId &async_request_id,
                                      size_t attempts, ProgressContext &prog_ctx,
                                      Context *on_finish) {
+  auto total_attempts = m_image_ctx.config.template get_val<uint64_t>(
+      "rbd_quiesce_notification_attempts");
+  if (total_attempts < attempts) {
```

Comment:
I don't understand the purpose of this `if` statement: why are we bothering to fetch `rbd_quiesce_notification_attempts` value on each invocation and then trying to adjust it to suit `attempts` counter?  If the idea is that `rbd_quiesce_notification_attempts` value might change in the interim, then we really aren't doing a good job here:
- if it gets decreased, e.g. from 10 to 1, then `1/10`, `1/9`, `1/8` would be logged for 1st, 2nd and 3rd attempt respectively
- if it gets increased, e.g. from 10 to 20, then `11/20`, `12/20`, `13/20` would be logged for 1st, 2nd and 3rd attempt respectively

Reply:
And, to make it worse, it's not just `ldout` log message that is affected but also `prog_ctx` progress tracker.

I would add `total_arguments` parameter to this method and thread through `on_notify` lambda the value that is passed in initially, from a 3-argument `notify_quiesce` overload.  Then it's always going to be correct.

Reply:
@idryomov Originally (before this patch) `total_attempts` was used only for `update_progress`, and as the configuration param may change at any time we could end up with weird "progress" values. That was why that `if (total_attempts < attempts)` condition. Now the same is applied for the log message.

Sure, we could just remember the initial value for `total_attempts`, e.g. pass it as an additional arg, though I am not sure if it is worth it though. Actually I am fine with any solution, even just ignore this unlikely issue and remove the `if` block.

Reply:
> @idryomov Originally (before this patch) `total_attempts` was used only for `update_progress`, and as the configuration param may change at any time we could end up with weird "progress" values. That was why that `if (total_attempts < attempts)` condition. Now the same is applied for the log message.

I don't see how that `if` condition helped avoid weird progress ratios even in the original code.  Calling `update_progress` with `0/9`, `0/8`, `0/7` on decrease to 1 or `11/20`, `12/20`, `13/20` on increase to 20 instead of the intended `1/10`, `2/10`, `3/10` is still weird in my book.

> 
> Sure, we could just remember the initial value for `total_attempts`, e.g. pass it as an additional arg, though I am not sure if it is worth it though. Actually I am fine with any solution, even just ignore this unlikely issue and remove the `if` block.

I think it's worth it.  If we are making an effort to both log locally and report progress over the wire, let's make it precise.

Reply:
> I don't understand the purpose of this if statement

The config value change when we are already in the process of sending attempts will not affect the infly request: when the request is started we take the number of attempts we want to try from the config and then just decrease this number after every attempt. It affect the progress value only though.

Consider a case, when an request started  the config value was 10, so `attempts` was set 10. Then after 2 unsuccessful attempts, when `attempts` became 8, the config value was changed to 5. It will still continue with remaining attempts, but the progress for this case (with `if` statement) will be:

  1/10, 2/10, 0/5, 0/5, 0/5 , 1/5, 2/5 ...
  
  It is not ideal, as it will look like the progress went back and temporarily got stuck, but it is definitely better than without `if` statement:
  
  1/10, 2/10, -2/5, -1/5, 0/5 , 1/5, 2/5 ...
  
Taking how unlikely such situation is for me the current solution looks good enough but sure I am fine with improving it.

Reply:
I hope the summary is to remove the `if` condition and preserve `total_attempts`, correct?

```
template <typename I>                                                           
void ImageWatcher<I>::notify_quiesce(const AsyncRequestId &async_request_id,
  ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆size_t attempts, ProgressContext &prog_ctx,
  ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆Context *on_finish) {
  static auto total_attempts = attempts;
  ceph_assert(attempts > 0);
  ldout(m_image_ctx.cct, 10) << this << " " << __func__ << ": async_request_id="
  ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆<< async_request_id << " attempts @ "
  ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆<< (total_attempts - attempts + 1) << "/"
  ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆ ┆<< total_attempts << dendl;
  auto notify_response = new watcher::NotifyResponse();
  auto on_notify = new LambdaContext(
  ┆ [notify_response=std::unique_ptr<watcher::NotifyResponse>(notify_response),
  ┆ ┆this, async_request_id, &prog_ctx, on_finish, attempts=attempts-1,
  ┆ ┆total_attempts](int r) {
  ┆ ┆ prog_ctx.update_progress(total_attempts - attempts, total_attempts);
  [...]
  }
```

Reply:
I don't think the above is correct: it would log `1/10`, `1/9`, `1/8` instead of `1/10`, `2/10`, `3/10`.  `total_attempts` should be passed into this method from the outside, not assigned from `attempts` inside.